src/tptp_parser.rs

```diff
@@ -181,7 +545,148 @@ pub fn parse_file<'a>(file: PathBuf) -> TPTPProblem {
+            }
+        }
+        Term::Function(n, ts) => {
+            if let Some(func) = func_map.get(&n) {
+                let func = func.clone();
+                let args = ts
+                    .into_iter()
+                    .map(|t| add_term_to_termbank(t, term_bank, var_map, func_map))
+                    .collect();
+                term_bank.mk_app(func, args)
+            } else {
+                let func = term_bank.add_function(FunctionInformation {
+                    name: n.to_string(),
+                    arity: ts.len(),
+                });
+                func_map.insert(n, func);
+                let args = ts
+                    .into_iter()
+                    .map(|t| add_term_to_termbank(t, term_bank, var_map, func_map))
+                    .collect();
+                term_bank.mk_app(func, args)
+            }
+        }
+    }
+}
+
+fn add_literal_to_termbank(
+    l: Literal,
+    term_bank: &mut TermBank,
+    var_map: &mut HashMap<Name, VariableIdentifier>,
+    func_map: &mut HashMap<Name, FunctionIdentifier>,
+) -> clause::Literal {
+    match l {
+        Literal::Eq(t1, t2) => {
+            let hash_cons_t1 = add_term_to_termbank(t1, term_bank, var_map, func_map);
+            let hash_cons_t2 = add_term_to_termbank(t2, term_bank, var_map, func_map);
+            clause::Literal::new(hash_cons_t1.clone(), hash_cons_t2.clone(), Polarity::Eq)
+        }
+        Literal::NotEq(t1, t2) => {
+            let hash_cons_t1 = add_term_to_termbank(t1, term_bank, var_map, func_map);
+            let hash_cons_t2 = add_term_to_termbank(t2, term_bank, var_map, func_map);
+            clause::Literal::new(hash_cons_t1.clone(), hash_cons_t2.clone(), Polarity::Ne)
+        }
+    }
+}
+
+fn to_literals(
+    t: SkolemTerm,
+    term_bank: &mut TermBank,
+    var_map: &mut HashMap<Name, VariableIdentifier>,
+    func_map: &mut HashMap<Name, FunctionIdentifier>,
+) -> Vec<clause::Literal> {
+    match t {
+        SkolemTerm::Literal(l) => vec![add_literal_to_termbank(l, term_bank, var_map, func_map)],
+        SkolemTerm::Or(s, t) => {
+            let mut literals_s = to_literals(*s, term_bank, var_map, func_map);
+            let mut literals_t = to_literals(*t, term_bank, var_map, func_map);
+            literals_s.append(&mut literals_t);
+            literals_s
+        }
+        SkolemTerm::And(_, _) => {
+            panic!("There should never be an 'And' inwards of an 'Or' at this stage!")
+        }
+    }
+}
+
+pub fn to_clauses(
+    t: SkolemTerm,
+    term_bank: &mut TermBank,
+    var_map: &mut HashMap<Name, VariableIdentifier>,
+    func_map: &mut HashMap<Name, FunctionIdentifier>,
+) -> Vec<Clause> {
+    match t {
+        SkolemTerm::Literal(l) => vec![Clause::new(vec![add_literal_to_termbank(
+            l, term_bank, var_map, func_map,
+        )])],
+        SkolemTerm::And(s, t) => {
+            let mut clause_s = to_clauses(*s, term_bank, var_map, func_map);
```

Comment:
In principle there is an approach here to accumulate everything into just one vec by threading one mutable reference through the recursive calls but we can safe that for later as well.